File: python/quick_sort.py
"""
    Quick sort.
"""


class QuickSort:
    def sort(self, l: list):
        self.qsort(l, 0, len(l)-1)

    def qsort(self, l: list, low: int, high: int):
        if low >= high: return
        m = self.partition(l, low, high)
        self.qsort(l, low, m-1)
        self.qsort(l, m+1, high)

    def partition(self, l: list, low: int, high: int):
        pivot, i = l[high], low
        for j in range(low, high):
            if l[j] < pivot:
                l[i], l[j] = l[j], l[i]
                i += 1
        l[i], l[high] = l[high], l[i]
        return i


if __name__ == '__main__':
    qs = QuickSort()
    ll = [4, 9, 1, 7, 3, 2, 8, 0, 6, 5]
    qs.sort(ll)
    print(ll)

File: golang/quick_sort.go
package main

import "fmt"

func main() {
	nums := []int{4, 3, 7, 8, 2, 5, 0, 1, 9, 6}
	sort(nums)
	fmt.Println(nums)
}

func sort(nums []int) {
	if len(nums) < 2 {
		return
	}
	qsort(nums, 0, len(nums)-1)
}

func qsort(nums []int, start, end int) {
	if start >= end {
		return
	}
	i := partition(nums, start, end)
	qsort(nums, start, i-1)
	qsort(nums, i+1, end)
}

func partition(nums []int, start, end int) int {
	pivot := nums[end]
	i := start
	for j := start; j < end; j++ {
		if nums[j] < pivot {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
	}
	nums[i], nums[end] = nums[end], nums[i]
	return i
}
